lib/IRGen/GenEnum.cpp

```diff
@@ -2940,11 +2940,32 @@ namespace {
         }
 
         case ForwardToPayload: {
-          auto &payloadTI = getPayloadTypeInfo();
-          // Apply the payload's operation.
-          addr = IGF.Builder.CreateElementBitCast(addr,
-                                                  payloadTI.getStorageType());
-          payloadTI.destroy(IGF, addr, getPayloadType(IGF.IGM, T), isOutlined);
+          auto emitPayloadDestroy = [&](const TypeInfo &payloadTI) {
+            // Apply the payload's operation.
+            addr = IGF.Builder.CreateElementBitCast(addr,
+                                                    payloadTI.getStorageType());
+            payloadTI.destroy(IGF, addr, getPayloadType(IGF.IGM, T),
+                              isOutlined);
+          };
+          if (TI->isCopyable(ResilienceExpansion::Maximal)) {
+            emitPayloadDestroy(getPayloadTypeInfo());
+            return;
+          }
+          auto &C = IGM.getLLVMContext();
+          llvm::BasicBlock *exitBlock = llvm::BasicBlock::Create(C);
+
+          llvm::BasicBlock *payloadBlock = llvm::BasicBlock::Create(C);
+
+          auto payload = ElementsWithPayload[0];
+
+          emitIndirectSwitch(IGF, T, addr, {{payload.decl, payloadBlock}},
+                             exitBlock, /*noLoad=*/false);
+
+          IGF.Builder.emitBlock(payloadBlock);
+          emitPayloadDestroy(*payload.ti);
+          IGF.Builder.CreateBr(exitBlock);
+
+          IGF.Builder.emitBlock(exitBlock);
```

Comment:
This is duplicating the destruction codegen we should already emit if we pick the `Normal` CopyDestroyStrategy. I think it would be better to choose the correct CopyDestroyStrategy to begin with.